From d1d9edff07a6c40886722f70c1f29e14e519378c Mon Sep 17 00:00:00 2001
From: Qi Zhang <qi1.zhang@intel.com>
Date: Wed, 25 Oct 2023 15:04:30 +0800
Subject: [PATCH] [IVI] Power key trigger forcesuspend instead of autosuspend

On Car, MCU controls IVI power flow and the suspend from
MCU to IVI is force suspend. Before the MCU is integrated,
we need some way to validate S3 capability from Android
to low level. So we change autosuspend to forcesuspend
when power key is pressed.

Tracked-On: OAM-112921
Signed-off-by: Qi Zhang <qi1.zhang@intel.com>
---
 .../android/server/power/PowerManagerService.java    | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/services/core/java/com/android/server/power/PowerManagerService.java b/services/core/java/com/android/server/power/PowerManagerService.java
index 29d773305e56..a9adf89c4e5e 100644
--- a/services/core/java/com/android/server/power/PowerManagerService.java
+++ b/services/core/java/com/android/server/power/PowerManagerService.java
@@ -3355,7 +3355,17 @@ public final class PowerManagerService extends SystemService
                         setHalInteractiveModeLocked(false);
                     }
                     if (!mDecoupleHalAutoSuspendModeFromDisplayConfig) {
-                        setHalAutoSuspendModeLocked(true);
+                        // setHalAutoSuspendModeLocked(true);
+                        mContext.enforceCallingOrSelfPermission(
+                            android.Manifest.permission.DEVICE_POWER, null);
+
+                        final int uid = Binder.getCallingUid();
+                        final long ident = Binder.clearCallingIdentity();
+                        try {
+                            forceSuspendInternal(uid);
+                        } finally {
+                            Binder.restoreCallingIdentity(ident);
+                        }
                     }
                 } else {
                     if (!mDecoupleHalAutoSuspendModeFromDisplayConfig) {
-- 
2.34.1

